File: sdk/go/messages.go
package wapi

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strconv"
	"time"
)

// Messages covers sending, reading and media. Session API key.
type Messages struct {
	t *transport

	// Media handles upload and inbound decryption.
	Media *Media
}

type Media struct{ t *transport }

// SendOption sets one field on a send. Which field you set decides what is sent — there is no
// separate route for images or groups — and setting two content fields is an error rather than a
// silent preference.
type SendOption func(map[string]any)

func Text(text string) SendOption      { return func(m map[string]any) { m["text"] = text } }
func ImageURL(url string) SendOption   { return func(m map[string]any) { m["imageUrl"] = url } }
func VideoURL(url string) SendOption   { return func(m map[string]any) { m["videoUrl"] = url } }
func AudioURL(url string) SendOption   { return func(m map[string]any) { m["audioUrl"] = url } }
func StickerURL(url string) SendOption { return func(m map[string]any) { m["stickerUrl"] = url } }

// DocumentURL sends a file. FileName is separate because the API takes it separately.
func DocumentURL(url, fileName string) SendOption {
	return func(m map[string]any) { m["documentUrl"], m["fileName"] = url, fileName }
}

// ReplyTo quotes an earlier message by its wapi msgId — not WhatsApp's string id.
func ReplyTo(msgID int64) SendOption { return func(m map[string]any) { m["replyTo"] = msgID } }

// Mentions tags participants in a group message.
func Mentions(numbers ...string) SendOption {
	return func(m map[string]any) { m["mentions"] = numbers }
}

// Field sets any other documented field — location, poll, contact — without the SDK having to
// model every shape the API accepts.
func Field(name string, value any) SendOption {
	return func(m map[string]any) { m[name] = value }
}

// Send sends a message. One endpoint for every type.
//
// "to" may be a phone number, a WhatsApp JID, or a group JID ending @g.us.
//
// A timeout here is ambiguous. It says the request failed, not that the message went undelivered —
// retrying blindly sends twice. Check UnavailableError.Ambiguous and reconcile with Info instead.
func (m *Messages) Send(ctx context.Context, to string, opts ...SendOption) (*SendResult, error) {
	body := map[string]any{"to": to}
	for _, opt := range opts {
		opt(body)
	}
	raw, err := m.t.do(ctx, "POST", "/api/send-message", nil, body)
	if err != nil {
		return nil, err
	}
	var out SendResult
	return &out, unwrap(raw, &out)
}

// Info fetches a sent message by its integer msgId.
//
// Returns WhatsApp's own record, so MessageTimestamp is a string and Status a number — see the
// MessageInfo doc comment.
func (m *Messages) Info(ctx context.Context, msgID int64) (*MessageInfo, error) {
	raw, err := m.t.do(ctx, "GET", "/api/messages/"+strconv.FormatInt(msgID, 10)+"/info", nil, nil)
	if err != nil {
		return nil, err
	}
	var out MessageInfo
	return &out, unwrap(raw, &out)
}

// Edit changes the text of a message you sent.
//
// WhatsApp allows this only for a short window afterwards and gives no way to ask how long is
// left, so a refusal is an ordinary outcome. The edit is a new message superseding the old one,
// so the response carries a fresh key alongside the original msgId.
func (m *Messages) Edit(ctx context.Context, msgID int64, text string) (json.RawMessage, error) {
	raw, err := m.t.do(ctx, "PUT", "/api/messages/"+strconv.FormatInt(msgID, 10), nil,
		map[string]any{"text": text})
	if err != nil {
		return nil, err
	}
	var out json.RawMessage
	return out, unwrap(raw, &out)
}

// Delete deletes a message for everyone. Same short window as editing.
//
// This endpoint puts message at the top level rather than under data, so it returns the
// confirmation string rather than unwrapping.
func (m *Messages) Delete(ctx context.Context, msgID int64) (string, error) {
	raw, err := m.t.do(ctx, "DELETE", "/api/messages/"+strconv.FormatInt(msgID, 10), nil, nil)
	if err != nil {
		return "", err
	}
	var out struct {
		Message string `json:"message"`
	}
	if err := json.Unmarshal(raw, &out); err != nil {
		return "", err
	}
	return out.Message, nil
}

// Resend retries a message whose status is "failed".
//
// Only failed messages, deliberately: a send that timed out is recorded as in_progress because
// nobody knows whether it arrived, and resending one of those is how a customer gets the same
// message twice.
func (m *Messages) Resend(ctx context.Context, msgID int64) (string, error) {
	raw, err := m.t.do(ctx, "POST", "/api/messages/"+strconv.FormatInt(msgID, 10)+"/resend", nil, nil)
	if err != nil {
		return "", err
	}
	var out struct {
		Message string `json:"message"`
	}
	if err := json.Unmarshal(raw, &out); err != nil {
		return "", err
	}
	return out.Message, nil
}

// MarkRead marks a received message as read.
//
// Takes the WhatsApp key rather than a msgId: inbound messages have no msgId, which only ever
// identifies something you sent. Take the key from the webhook payload.
func (m *Messages) MarkRead(ctx context.Context, key MessageKey) error {
	_, err := m.t.do(ctx, "POST", "/api/messages/read", nil, map[string]any{"key": key})
	return err
}

// React reacts to a message.
//
// A wapi extension, not part of the WasenderAPI interface — they report reactions over webhooks
// but offer no way to send one. Feature-detect if you target both.
func (m *Messages) React(ctx context.Context, key MessageKey, emoji string) error {
	_, err := m.t.do(ctx, "POST", "/api/messages/react", nil, map[string]any{
		"key": key, "emoji": emoji,
	})
	return err
}

// Unreact removes a reaction. An empty emoji is WhatsApp's convention, not a separate endpoint.
func (m *Messages) Unreact(ctx context.Context, key MessageKey) error {
	return m.React(ctx, key, "")
}

// Upload stores bytes and returns a permanent URL to pass to Send.
//
// Permanent is the point: media is fetched server-side at send time, so an expiring link would
// stop working between upload and delivery. Caps at 16 MB.
//
// publicUrl sits at the top level, not under "data".
func (md *Media) Upload(ctx context.Context, content []byte, mimetype, fileName string) (string, error) {
	body := map[string]any{
		"base64":   base64.StdEncoding.EncodeToString(content),
		"mimetype": mimetype,
	}
	if fileName != "" {
		body["fileName"] = fileName
	}
	// Uploads are slower than reads; the client default is too tight for a large file.
	ctx, cancel := context.WithTimeout(ctx, 2*time.Minute)
	defer cancel()

	raw, err := md.t.do(ctx, "POST", "/api/upload", nil, body)
	if err != nil {
		return "", err
	}
	var out struct {
		PublicURL string `json:"publicUrl"`
	}
	if err := json.Unmarshal(raw, &out); err != nil {
		return "", fmt.Errorf("wapi: decoding response: %w", err)
	}
	return out.PublicURL, nil
}

// Decrypt turns an inbound encrypted media node into a URL valid for one hour.
//
// Inbound media arrives as a CDN link plus a mediaKey; the bytes are useless without decryption,
// and only the session holding the keys can do it. Pass the message object from the webhook.
func (md *Media) Decrypt(ctx context.Context, message json.RawMessage) (string, error) {
	body := map[string]any{"data": map[string]any{"messages": map[string]any{"message": message}}}
	raw, err := md.t.do(ctx, "POST", "/api/decrypt-media", nil, body)
	if err != nil {
		return "", err
	}
	var out struct {
		PublicURL string `json:"publicUrl"`
	}
	if err := json.Unmarshal(raw, &out); err != nil {
		return "", fmt.Errorf("wapi: decoding response: %w", err)
	}
	return out.PublicURL, nil
}

File: sdk/python/wapi/resources/messages.py
"""Sending, reading and media. Session API key."""

from __future__ import annotations

import base64 as _base64
from typing import Any

from .._http import Transport, data


class Media:
    def __init__(self, http: Transport) -> None:
        self._http = http

    def upload(self, content: bytes, mimetype: str, file_name: str | None = None) -> str:
        """Upload bytes and get a **permanent** URL to pass to :meth:`MessagesResource.send`.

        Permanent is the point: media is fetched server-side at send time, so an expiring link
        would stop working between upload and delivery.

        ``publicUrl`` sits at the top level, not under ``data``. Caps at 16 MB.
        """
        body = {
            "base64": _base64.b64encode(content).decode("ascii"),
            "mimetype": mimetype,
        }
        if file_name:
            body["fileName"] = file_name
        # Uploads are slower than reads; the default deadline is too tight for a large file.
        return self._http.request("POST", "/api/upload", body=body, timeout=120.0)["publicUrl"]

    def decrypt(self, message: dict[str, Any]) -> str:
        """Turn an inbound encrypted media node into a URL valid for one hour.

        Inbound media arrives as a CDN link plus a ``mediaKey``; the bytes are useless without
        decryption, and only the session holding the keys can do it. Pass the ``message`` object
        straight from the webhook payload.
        """
        return self._http.request(
            "POST", "/api/decrypt-media", body={"data": {"messages": {"message": message}}}
        )["publicUrl"]


class MessagesResource:
    def __init__(self, http: Transport) -> None:
        self._http = http
        self.media = Media(http)

    def send(self, to: str, **content: Any) -> Any:
        """Send a message. One endpoint for every type.

        Which field you set decides what is sent — ``text``, ``imageUrl``, ``documentUrl``,
        ``location``, ``poll`` and so on. Setting two content fields is an error rather than a
        silent preference. ``to`` may be a phone number, a JID, or a group JID ending ``@g.us``.

        **A timeout here is ambiguous.** It says the request failed, not that the message went
        undelivered — retrying blindly sends twice. Reconcile with :meth:`info` instead.
        """
        return data(self._http.request("POST", "/api/send-message", body={"to": to, **content}))

    def info(self, msg_id: int) -> Any:
        """Fetch a sent message by its integer ``msgId``.

        Two fields do not match what :meth:`send` returns, because this route returns WhatsApp's
        own record: ``messageTimestamp`` is a **string** (a protobuf int64) and ``status`` is a
        **number** — 0 error, 1 pending, 2 sent, 3 delivered, 4 read.
        """
        return data(self._http.request("GET", f"/api/messages/{msg_id}/info"))

    def edit(self, msg_id: int, text: str) -> Any:
        """Edit a message you sent.

        WhatsApp allows this only for a short window afterwards and gives no way to ask how long
        is left, so a refusal is an ordinary outcome. The edit is a *new* message superseding the
        old one, so the response carries a fresh key alongside the original ``msgId``.
        """
        return data(self._http.request("PUT", f"/api/messages/{msg_id}", body={"text": text}))

    def delete(self, msg_id: int) -> str:
        """Delete a message for everyone. Same short window as editing.

        This endpoint puts ``message`` at the *top level* rather than under ``data``, so it
        returns the confirmation string rather than unwrapping.
        """
        return str(self._http.request("DELETE", f"/api/messages/{msg_id}")["message"])

    def resend(self, msg_id: int) -> str:
        """Retry a message whose status is ``failed``.

        Only failed messages, deliberately: a send that timed out is recorded as ``in_progress``
        because nobody knows whether it arrived, and resending one of those is how a customer
        gets the same message twice.
        """
        return str(self._http.request("POST", f"/api/messages/{msg_id}/resend")["message"])

    def mark_read(self, key: dict[str, Any]) -> Any:
        """Mark a *received* message as read.

        Takes the WhatsApp ``key`` rather than a ``msgId``: inbound messages have no ``msgId``,
        which only ever identifies something you sent. Take the key from the webhook payload.
        """
        return data(self._http.request("POST", "/api/messages/read", body={"key": key}))

    def react(self, key: dict[str, Any], emoji: str) -> Any:
        """React to a message.

        **A wapi extension**, not part of the WasenderAPI interface — they report reactions over
        webhooks but offer no way to send one. Feature-detect if you target both.

        Keyed like :meth:`mark_read`, and for the same reason.
        """
        return data(
            self._http.request("POST", "/api/messages/react", body={"key": key, "emoji": emoji})
        )

    def unreact(self, key: dict[str, Any]) -> Any:
        """Remove a reaction. An empty emoji is WhatsApp's convention, not a separate endpoint."""
        return self.react(key, "")
